sim/prng.js

```diff
@@ -96,6 +96,35 @@ class PRNG {
 		return this.next(denominator) < numerator;
 	}
 
+	/**
+	 * Return a random item from the given array.
+	 *
+	 * This function chooses items in the array with equal probability.
+	 *
+	 * If there are duplicate items in the array, each duplicate is
+	 * considered separately. For example, sample(['x', 'x', 'y']) returns
+	 * 'x' 67% of the time and 'y' 33% of the time.
+	 *
+	 * The array must contain at least one item.
+	 *
+	 * The array must not be sparse.
+	 *
+	 * @param {ReadonlyArray<T>} items - the items to choose from
+	 * @return {T} - a random item from items
+	 * @template T
+	 */
+	sample(items) {
+		if (items.length === 0) {
+			throw new RangeError(`Cannot sample an empty array`);
+		}
+		const index = this.next(items.length);
+		const item = items[index];
+		if (item === undefined && !Object.prototype.hasOwnProperty.call(items, index)) {
+			throw new RangeError(`Cannot sample a sparse array`);
```

Comment:
I assume Slayer95 just means he disapproves. I agree with him, this feels wrong.

On one hand, why _can't_ we sample a sparse array? Maybe I _want_ to usually get `undefined` back if I sample a sparse array? That seems like the expected behavior.

On the other hand, I dislike this for the same reason I dislike `hasOwnProperty` calls in `for...in` loops. Even if the perf issue is unnoticeable, the readability issue is not great. The general expectation that the code should account for bad code sends the wrong message. Like using workarounds for `undefined` being overwritten in older browsers. We don't want people to go out of their way to account for buggy coding styles not used in PS at all.

Reply:
I think I'd be ok with this kind of check in tests (although I'd still consider it kinda useless), but not in the function implementation proper.

Reply:
> On one hand, why *can't* we sample a sparse array? Maybe I *want* to usually get undefined back if I sample a sparse array? That seems like the expected behavior.

Expected by who? =]

The type signature of `randomItem` says the function does *not* return `undefined` (unless the input array contained `undefined`). Looking at the types, returning `undefined` is *unexpected*.

Some standard library functions, such as `Array.prototype.forEach`, do *not* look at missing array elements. If you're familiar with how these functions behave of sparse arrays, you might be surprised by `randomItem` returning `undefined`:

```js
const items = [];
items[10] = 42;
items.forEach((item) => console.log(item));
// => 42
```

I think both expectations (missing items are sampled; missing items are not sampled) are valid. Because it's impossible to satisfy both expectations at the same time, I think the implemented behaviour should be explicit (`randomItemByIndex`/`randomItemOrSparseHole`; `randomItemIgnoringSparseHoles`) or disallowed (using the type system, or by reporting an error if `randomItem` is called with a sparse array).

> The general expectation that the code should account for bad code sends the wrong message.
 Like using workarounds for `undefined` being overwritten in older browsers.

The check *stops* bad code. The check doesn't *work around* bad code. I think your comparison with `undefined` is unfair, because re-defining `undefined` *works around* bad code.

Reply:
I would consider a sparse array to contain `undefined`, but I recognize it's possibly controversial. In practice it doesn't matter because all sparse arrays should be `Map`s.

I still don't like the idea of checking for a condition that will never happen, but I guess we do it in other cases so if you care this strongly I guess I don't mind leaving the check in.